### DISCUSSION TITLE
IBC Expansion

### DISCUSSION THREAD

**Discussion by adiraviraj** [cosmos/ibc-go#5095] [General]
### Introduction
As of today, ibc-go is used by ~100 Cosmos SDK chains. IBC has clearly achieved PMF within Cosmos but needs to expand its extensibility to different ecosystems. 

To realize **IBC’s vision of becoming the TCP/IP for blockchains**, we need a deliberate and strategic approach to extend integrations with other chains.

### Objective
This discussion aims to:
1. Present chains and rollup frameworks written in Go for potential IBC integrations using a generic ibc-go module.
2. Gather feedback from our users to understand which chains/frameworks have the highest demand for integrations. 

### Current Efforts
Apart from Cosmos, IBC is currently live on Polkadot. Ongoing efforts to integrate IBC into other ecosystems include:

- [Polymer Labs:](https://www.polymerlabs.org/) Cosmos<>Ethereum mainnet
- [Landslide Network:](https://www.landslide.network/) Cosmos<>Avalanche
- [Composable Finance:](https://www.composable.finance/) Cosmos<>NEAR, Cosmos<>Solana, Cosmos<>Ethereum mainnet
- [Union:](https://union.build/) Cosmos<>Ethereum mainnet
- [Electron Labs:](https://electronlabs.org/) Cosmos<>Ethereum mainnet
- [Datachain:](https://www.datachain.jp/) IBC enterprise adoption

While these contributions play a crucial role in driving IBC’s extensibility, the focus of this discussion is on the pathway for ibc-go expansion into other chains/frameworks.

### Potential targets
In addition to the above-mentioned efforts to expand IBC, Interchain GmbH is actively exploring additional Go-based chains and frameworks. Our initial research has identified the following list of potential targets:

- [OP Stack:](https://stack.optimism.io/) Framework to build optimistic (or ZK rollups in the future) on Ethereum*.
- [Arbitrum Nitro:](https://github.com/OffchainLabs/nitro) Tech stack used for the [Arbitrum One](https://arbitrum.io/rollup) rollup.
- [Rollkit:](https://github.com/rollkit/rollkit) Framework to build rollups using the Cosmos SDK. Switches out CometBFT for a single sequencer or a shared sequencer network. Plugs into arbitrary DA layers. 
- [Polygon:](https://github.com/maticnetwork/) The [Polygon PoS](https://github.com/maticnetwork/bor) chain uses a fork of the Cosmos SDK and Tendermint. But the chain will [migrate to a ZK validium soon](https://forum.polygon.technology/t/pre-pip-discussion-upgrading-polygon-pos-to-a-zkevm-validium/12187). The existing [Polygon zkEVM rollup](https://github.com/0xPolygonHermez/zkevm-node) and/or the [Polygon CDK](https://github.com/0xPolygon/cdk-validium-node) are potential integration targets, but more investigation is needed to understand the technical feasibility for an integration.
- [Filecoin:](https://github.com/filecoin-project/lotus) L1 chain utilizing IPFS for data storage and retrieval
- [Celo:](https://github.com/celo-org/celo-blockchain) An EVM-compatible L1 chain. [Currently migrating to an OP Stack rollup](https://forum.celo.org/t/clabs-proposal-for-celo-to-transition-to-an-ethereum-l2/6109).
- [Algorand:](https://github.com/algorand/go-algorand) General-purpose L1 
- [MultiversX (formerly Elrond):](https://github.com/multiversx/mx-chain-go) General-purpose L1

*_An OP stack integration would also onboard [OP Mainnet](https://app.optimism.io/bridge/deposit), [Base](https://base.org/), and [BNB OP chain](https://opbnb.bnbchain.org/en), as well as existing L1s that are migrating to an op-rollup such as [Celo](https://celo.org/)._

<br>The criteria for prioritizing the aforementioned chains/frameworks include:

- **Developer Interest:** We consider where developers are currently most interested in building and how this interest may evolve as chains/frameworks mature.
- **On-Chain Activity:** We assess which ecosystems currently have the highest market share, unique Daily Active Users (DAU), and where high-value apps are deployed.
- **Technical Feasibility:** We evaluate whether the chain/framework is written in Go and if there are any specific architectural constraints that could pose challenges for an ibc-go integration, given our available resources.
- **Market Trends:** We take into account industry trends and examine how this landscape is evolving. This helps us anticipate which frameworks are likely to attract the majority of developer interest and therefore, user demand.

These criteria guide our decision-making process to strategically prioritize and select the most useful chains/frameworks for integration.

Within our team, we employ the [ICE framework](https://itamargilad.com/the-tool-that-will-help-you-choose-better-product-ideas/) for prioritization decisions. In the context of external integrations, the framework could be applied as follows to arrive at a prioritization score:
- **Impact:** Assessing the positive effect (onboarding more developers, increased cross-chain volume, etc.) of a specific integration on IBC users.
- **Confidence:** Gauging our certainty regarding the positive impact of a particular integration.
- **Ease:** Evaluating the effort, time, and resources required for the integration.

### Community feedback
In the spirit of open-source development, we’d like to open up this discussion to the public to ensure that the IBC community as a whole is aligned on integrations with the highest potential impact. We’re looking for feedback on:
- **User Demand:** As a user of ibc-go, which ecosystem would you like ibc-go to integrate with and why?
- **Strategic Value:** Apart from the target ecosystems listed above, what are other potential integrations that can provide the highest strategic value to IBC?

Your input is highly valuable in shaping the future of IBC integration. We are committed to solving user problems, and your feedback is instrumental in ensuring that our efforts align with the real needs of our users.

### Action plan
Given our limited engineering resources, our goal is to strategically focus on 2-3 ecosystems. A high-level action plan on how to execute new integrations could look as follows:

1. Use the insights and feedback collected from this discussion, along with our internal research findings around feasibility, architectural constraints, and market trends to narrow down a concise list of 2-3 target ecosystems for ibc-go integration.

2. Create simple POCs for each selected ecosystem. This could include minimal tasks such as performing light client updates or negotiating a channel handshake. The purpose of these POCs is to gain practical experience with each ecosystem and to identify which changes are generic and which changes may necessitate more comprehensive design considerations.

3. Building on the insights and findings from the POC phase, we’ll draft RFCs, outlining the specs, design decisions, and requirements for integrations with each ecosystem.

4. With the RFCs as a guide, we’ll move on to implementing more advanced POCs. These may encompass relatively more complex tasks such as token transfers.

5. Begin implementation work based on the specifications and design considerations outlined in the RFC.

6. Initiate release cycle

7. Release

Your feedback is important in helping us determine which chains/frameworks to prioritize, ensuring that our approach aligns with the needs of the IBC community. We invite you to share your thoughts and suggestions on potential integrations. Thank you for your participation!

**Comment by zmanian** [comment]
Some additional context

Polygon CDK chains should be able to validate wasm blobs as a result of the Near/Polygon zkwasm collaboration

**Comment by adizere** [comment]
Great write-up!

I have two elements of critical feedback on the process, see below. 
I'll add later thoughts about feedback on user demand and strategic value.

> Given our limited engineering resources, our goal is to strategically focus on 2-3 ecosystems

A firm decision on 1, and executing on that 1 integration, fast and smooth, is likely better than focusing on 2-3.

> Ease: Evaluating the effort, time, and resources required for the integration.

I think the list of potential targets is a really good starting point. To help get an accurate evaluation of the "Ease" of integration, I would also rank (or filter) the potential targets by how likely is it that the said ecosystems are interested in IBC. I am mentioning this because it's possible a couple of the "potential targets" would not be very receptive of having IBC integrated, since they have their ideas about solutions in the interoperability space; even if those ideas could coexist with IBC, it's usually an uphill battle. Briefly, which of the "potential targets" have appetite for IBC and are open to collaborating/feedback with Interchain GmbH (and other integrators)?

**Reply by adiraviraj** [reply]
thanks for the feedback! i really appreciate it.

re the two elements that you mentioned:

> A firm decision on 1, and executing on that 1 integration, fast and smooth, is likely better than focusing on 2-3.

completely agree. the reason we'd like to investigate 2-3 targets is so that we have a better understanding of what design decisions are needed to make ibc-go generic for these targets. in practice, given the resources at hand, it's likely that we might pivot to prioritizing 1 target.

> I think the list of potential targets is a really good starting point. To help get an accurate evaluation of the "Ease" of integration, I would also rank (or filter) the potential targets by how likely is it that the said ecosystems are interested in IBC. I am mentioning this because it's possible a couple of the "potential targets" would not be very receptive of having IBC integrated, since they have their ideas about solutions in the interoperability space; even if those ideas could coexist with IBC, it's usually an uphill battle. Briefly, which of the "potential targets" have appetite for IBC and are open to collaborating/feedback with Interchain GmbH (and other integrators)?

this is a good point. as of now, the OP Stack team has been receptive to the idea of integrating IBC. and we've been working closely with the team at Rollkit to have IBC on their stack.

**Comment by 0xBrainjar2** [comment]
One piece of feedback from my perspective is that I believe that ibc-go should also maintain the contracts that go into the wasm client. As we trend to connecting multiple ecosystems over IBC, the wasm client will continue to be an important piece of infrastructure, and Composable is interested in collaborating to ensure the client contracts going into the wasm client are fully up to date etc.

**Comment by hxrts** [comment]
Rollkit and then the OP stack seems like the most feasible path for IBC go integration targets. Rollkit should be pretty straightforward and give the development team hands-on experience with a fraud proof architecture, and then the OP stack is also written in go, has people building app chains, high profile users, and Tess is about as good a counterparty as we can ask for to help integrate on the OP side.

**Reply by hxrts** [reply]
note, it may be best to work with polymer on the OP integration since they recently decided to become an OP chain. so maybe there's a separation of concerns where the ICF team could convert IBC go to use [ADR-05](https://github.com/cosmos/ibc-rs/blob/main/docs/architecture/adr-005-handlers-redesign.md) and have have polymer implement machinery specific to OP

**Reply by notbdu** [reply]
We're planning on supporting cosmos SDK <> OP stack rollups. IBC-go would come in working form as a cosmos sdk module. The lift would be on the client side. For example, connectivity between a cosmos sdk eth rollup to a cosmos sdk app chain would require the app chain to run an eth LC.

**Comment by Simonch0o** [comment]
Arbitrum and OP stack would definitely be strong moves, I think Solana may also be a strong potential choice If it's possible. 
There's a lot of chains listed on the target list that don't really have the users and liquidity and I believe we should focus on a few key ones to start.

Ton may be an interesting one to follow in the future too, as it opens up a very different market, but may be totally different.

**Comment by storiumi** [comment]
If we focus on the enterprise adoption (as mentioned in current efforts), we are hearing some enterprises employ Polygon for their platform, probably more often comparing to other choices. (Please regard this opinion as observation, not suggestion.)
Please bear in mind to consider accepting flexibility in encoding when expanding IBC(-App) to other ecosystems as JSON is not always the most efficient way. (As discussed in https://forum.openibc.com/t/roast-ibc-transport-tao-primitives/19/9 )

**Comment by nashqueue** [comment]
Rollkit will be the easiest to integrate as it shares common cometBFT types that IBC uses and the cosmos-sdk. The Progression could be the following : 
1. JTMB IBC (Just Trust Me Bro, single sequencer verifications )
2. JTMB IBC + verifying inclusion proofs using conditional clients with a DA-Client ( This will allow IBC chains to post to a DA-Layer and verify inclusion having an on-chain DA-Layer Light Client) 
3. Optimistic IBC 
4. Iterate

The learnings and concepts learned about Rollup bridging can then be used to tackle further frameworks.

**Comment by oct-sheldon** [comment]
Octopus Network is worth a mention: Composable has NEAR IBC on their roadmap but the Octopus Network team has already completed an audit of the same materials (further drafts still going forward for improvements) without the requirement of a broker/orchestrator chain. Sorry to be late

https://github.com/octopus-network/near-ibc

https://medium.com/oct-network/introduction-to-near-ibc-how-to-implement-the-ibc-protocol-with-smart-contracts-a8b0e13cb886

**Comment by flippingalgos** [comment]
Just circling back on this. Any status on adding support for algorand??? If not what can i do to make this happen?

**Reply by adiraviraj** [reply]
we currently have no roadmap item to support algorand. are you someone from the algorand team?

**Reply by johnalanwoods** [reply]
I’m the CTO, you should prioritise if possible, we have a bustling ecosystem.

**Reply by womensrights** [reply]
I've sent you a message on linkedin John to make this happen so let's take it from there!
